Burnman profiles

gassmoeller opened the pull request:
This is work in progress towards creating a model that uses a fully consistent (TM) equation of state created by the BurnMan software.

bangerth commented:
Let me know once you want me to look at this. You could also break out the various ascii data classes for separate review.

jdannberg commented:
@bangerth I think you can have a first look at this now. We may want to extend the cookbook a bit (to compare models with the different formulations), but that shouldn't change anything else.

tjhei commented on source/gravity_model/ascii_data.cc:
```diff
@@ -0,0 +1,118 @@
+      Point<dim> vertical_direction;
+
+      // in dependence of what the geometry model is, gravity points in a different direction
+      if (const GeometryModel::SphericalShell<dim> *geometry = dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model()))
```
I wonder if there is a better way to do this. Maybe we can introduce a property that provides a "downward direction" for a point x? At least you can write this as something like
```
if (dynamic_cast<const GeometryModel::SphericalShell<dim>*> (&this->get_geometry_model())
|| dynamic_cast<const GeometryModel::Chunk<dim>*> (&this->get_geometry_model())
|| ... )
return - magnitude * position/position.norm();
else
...
```
(note that you don't use the ``gm`` variable)

tjhei commented on source/material_model/non_dimensional.cc:
```diff
@@ -0,0 +1,194 @@
+/*
+  Copyright (C) 2016 by the authors of the ASPECT code.
```
please remove the nondimensional model (and rebase to current master)

tjhei commented on data/adiabatic-conditions/ascii-data/simple_test.txt:
```diff
@@ -0,0 +1,6 @@
+# Only next line is parsed in format: [nx] because of keyword "POINTS:"
+# POINTS: 2
+# Columns: depth pressure temperature density gravity
+0.       1. 3. 5. 7.
```
this file has different columns then the file above, is that on purpose?

bangerth commented on source/gravity_model/ascii_data.cc:
```diff
@@ -0,0 +1,114 @@
+                                  "ascii data",
+                                  "Gravity is read from an ascii data file "
+                                  "that contains a vertical or radial profile. "
+                                  "File structure : ... TBD")
```
to be finished :-)

jdannberg replied:
@gassmoeller Can you add the file structure for all of the new plugins?

bangerth commented on source/material_model/ascii_reference_profile.cc:
```diff
@@ -0,0 +1,212 @@
+    ASPECT_REGISTER_MATERIAL_MODEL(AsciiReferenceProfile,
+                                   "ascii reference profile",
+                                   "A material model that reads in a reference "
+                                   "state from a text file.")
```
I think this description should include (i) a discussion of the file format, (ii) a representation of the formulas that you use in the `evaluate()` function so that one can match the parameters to the actual model.

bangerth commented on source/utilities.cc:
```diff
@@ -1549,6 +1549,43 @@ namespace aspect
+      const std::string filename = Utilities::AsciiDataBase<dim>::data_directory
+                                   + Utilities::AsciiDataBase<dim>::data_file_name;
+
+      if (fexists(filename))
```
Use `Utilities::fexists` instead.

bangerth commented on tests/ascii_data_adiabatic_conditions.prm:
```diff
@@ -0,0 +1,84 @@
+  set Model name = ascii data
+  subsection Ascii data model
+    set Data directory       = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
+    set Data file name       = simple_test.txt
```
As mentioned above, I think it would be best to just move this file into the `tests/` directory, unless it is of any use to users.

jdannberg commented on data/adiabatic-conditions/ascii-data/simple_test.txt:
```diff
@@ -0,0 +1,7 @@
+# Test data for ascii data reference state
```
It is only used for a test, but at the moment all of the ASCII data plugins have their test data in the data directory (so I just wanted to do it in a consistent way). Do we want to move all of them? They can also be seen as example data so that users have a very simple example file where they see the data structure, and the data directory is where you would normally look for something like this.

tjhei replied:
I was also surprised to see test data in /data/. Because the data is not at all sensible, I would would to move this into tests. I can make a PR for the other files.

bangerth replied:
That would be my preference indeed.

jdannberg commented on doc/manual/cookbooks/burnman/plot_reference_profile.plt:
```diff
@@ -0,0 +1,55 @@
+set rmargin 1
+
+set title "Reference temperature"
+plot "../../../../data/adiabatic-conditions/ascii-data/isentrope_properties.txt" using 3:($1/1000) with lines linecolor rgb "gray" lw 4 lt 1 title "temperature"
```
That would have been due to the latent heat, but we decided to run the model without latent heat effects (and the updated profile does not have the drop any more).

bangerth replied:
OK, great.

jdannberg commented on doc/manual/manual.tex:
```diff
@@ -6795,8 +6795,8 @@ \subsubsection{3D convection with an Earth-like initial condition}
-\marginpar{Use cookbooks/periodic-box.prm}
-\marginpar{Finish the GPlates section}
+%\marginpar{Use cookbooks/periodic-box.prm}
+%\marginpar{Finish the GPlates section}
```
I have actually no idea if these lines are still relevant (one is yours, one is Rene's), I just didn't like the point that there are these TODOs in the manual (and now that I am looking, I found a couple of marginpars like these ones). We could put them in a different place (for example as issues on github), or maybe there's a better way, like opening a project on github and than adding all the manual fixes we want to do as notes, but I don't think we should have marginpars in the manual (because that's a also not the place where I would look if I had time for half an hour of free time and wanted to fix something in ASPECT, and also because we will forget to remove them for the releases).

bangerth replied:
Yes, there are a couple of marginpars that have survived essentially the entire history of ASPECT. We did not have an issue tracker in the beginning that would have allowed us to track notes such as these outside the manual. They should not be in the manual any more these days.

Feel free to remove these two -- I think they really are resolved -- and remove others in a separate patch if you open an issue for each marginpar that is still relevant. Just don't comment them out -- that doesn't help anyone :-)

jdannberg replied:
Okay, sure, I'll do that.

gassmoeller commented on doc/manual/manual.tex:
```diff
@@ -7008,6 +7008,104 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+If we look at the output of our model (for example in \texttt{paraview}), we can see how cold, highly 
+viscous slabs are subducted and hot plumes rise from the core-mantle boundary. The final time step of 
+the model is shown in Figure~\ref{fig:burnman-convection}, and the full model evolution can be found 
+at \url{https://youtu.be/FS-ru4N0p-c}.
```
No. But it would be great to have access, I have some more videos, and also a few minor things for the website.

bangerth replied:
I don't think I can give you access (they're hosted in Heidelberg, and while I can check things in I don't have the privileges to create accounts there), but if you want to edit something, pull it from the website, edit it, and send it to me -- I'll check it in for you.

jdannberg commented on include/aspect/utilities.h:
```diff
@@ -691,6 +691,43 @@ namespace aspect
+         * Pointer to an object that reads and processes data we get from text
+         * files.
+         */
+        std_cxx11::shared_ptr<aspect::Utilities::AsciiDataLookup<1> > lookup;
```
Rene, can you have a look at this? I think we just copied it from `AsciiDataInitial`...

bangerth replied:
I also don't quite recall whether we had `unique_ptr` already in 8.4....

gassmoeller commented:
This PR should be ready for a second round of review. We addressed all points of the previous review, and following the discussion with @tjhei the ascii data reader can now optionally provide names for the data columns (following the gnuplot format of naming data columns), and plugins can require the data file to contain certain columns. So far only the new adiabatic conditions, gravity, and material model plugins require these names, any existing ascii data plugins still use the hard-coded order of columns (this is something for another day, and it is also not really necessary for the other plugins, it is just convenient to have one reference state file for several plugins).

tjhei commented:
This looks good and I like the new implementation to identify columns by names!

I wonder if we should give these text files a different extension than .txt (maybe .asciitable or something) to make it obvious that there are some restrictions to the data format?

gassmoeller commented:
Maybe, although .txt makes it immediately clear that it is human readable. Anyway, if we want to rename the data files, we should probably rename all at once and in a separate PR?

tjhei commented:
> Anyway, if we want to rename the data files, we should probably rename all at once and in a separate PR?

It would fit with the changes you are doing in the AsciiData classes, but the PR is getting quite big already (it is already addressing several independent things!). So, I am fine with it either way.

gassmoeller commented:
So this is ready for merge? Or would you like to take another look @bangerth? I would like to give this a few tests before the release, so the earlier this is merged the better :smile:

bangerth commented on cookbooks/burnman.prm:
```diff
@@ -0,0 +1,140 @@
+
+set Dimension                              = 2
+set Use years in output instead of seconds = true
+set End time                               = 2.63e8
```
Interesting choice -- but then, why not :-)

bangerth commented on cookbooks/burnman.prm:
```diff
@@ -0,0 +1,140 @@
+
+  subsection Ascii data model
+    set Data file name = isentrope_properties.txt
+    set Data directory = $ASPECT_SOURCE_DIR/data/adiabatic-conditions/ascii-data/
```
since the file name is relative to the directory, maybe reverse the order of parameters here? That is also what you do further down

bangerth commented on data/adiabatic-conditions/ascii-data/isentrope_properties.txt:
```diff
@@ -0,0 +1,504 @@
+# POINTS: 501 
```
There can presumably be many isentropic adiabatic profiles, depending on which material model is used. As a consequence, would it make sense to reflect this in the name of the file? E.g., Burch-Monaghan-isentrope-properties.txt or something?

tjhei replied:
It is "Birch-Murnaghan" - you had me laughing out loud on this one.

While I agree that "isentrope" is not a great name, the equation of state is not the only parameter either of course. I see this as one of possibly many example profiles.

bangerth replied:
Close enough with "Burch-Monaghan" :-)

bangerth commented on doc/manual/manual.tex:
```diff
@@ -1939,6 +1940,8 @@ \subsubsection{Combined formulations}
 \end{itemize}
 
+An example model that shows a comparison between different approximations is shown in the 
+cookbook~\ref{sec:cookbooks-burnman}.
```
-> cookbook shown in Section~\ref...

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+the (truncated) anelastic liquid approximation (see Sections~\ref{sec:ala} and 
+\ref{sec:tala}), together with a reference profile read in from an ASCII data file.
+The data we use here is generated with the open source mineral physics toolkit BurnMan 
+(\url{www.burnman.org}) using the python example program \texttt{simple\_adiabat.py}. 
```
http://www.burnman.org/

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+The viscosity in this model is computed as the product of a radial profile $\eta_r(z)$ and a 
+term that describes the dependence on temperature:
+\begin{align*}
+\eta(z,T) = \eta_r(z) \eta_0 \exp\left(-A \frac{T - T_\text{adi}}{T_\text{adi}}\right),
```
Since we work in Cartesian coordinates, the use of `z` may be confusing without explanation.

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+\texttt{Viscosity} and \texttt{Thermal viscosity exponent}, and $\eta_r(z)$ is a stepwise constant 
+function that determines the viscosity profile. This function can be specified by providing a list of 
+\texttt{Viscosity prefactors} and a list of depths that describe in which depth range each prefactor 
+should be applied, in other words, in which depth the viscosity changes. By default, it is set
```
-> AT which depth...

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+function that determines the viscosity profile. This function can be specified by providing a list of 
+\texttt{Viscosity prefactors} and a list of depths that describe in which depth range each prefactor 
+should be applied, in other words, in which depth the viscosity changes. By default, it is set
+to viscosity jumps in 150\,km depth, between upper mantle and transition zone, 
```
in -> at

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+which in the limit of the anelastic liquid approximation becomes Equation~\eqref{eq:stokes-ALA-1}). 
+In addition, the density derivative in the mass conservation equation 
+(see Section~\ref{sec:mass-conservation-approximation}) is taken from the adiabatic 
+conditions, where it is computed as depth derivative of the provided reference density profile
```
-> as THE depth

bangerth commented on doc/manual/manual.tex:
```diff
@@ -7021,6 +7021,158 @@ \subsubsection{Using reconstructed surface velocities by GPlates}
+  \includegraphics[width=0.48\textwidth]{cookbooks/burnman/viscosity}
+  \caption{\it Compressible convection in a 2d spherical shell, using a reference profile exported 
+               form BurnMan, which is based on the Birch-Murnaghan equation of state. Shown is the
+               state at the end of the model evolution over 260\,Ma.}
```
In my zeal to eradicate the passive voice, let me suggest to change the last sentence to "The figure shows the state at the end....".

bangerth commented on include/aspect/utilities.h:
```diff
@@ -379,21 +379,41 @@ namespace aspect
      * needs to contain the number of grid points in each dimension as for
-     * example '# POINTS: 3 3'. The order of the columns has to be
+     * example '# POINTS: 3 3'. The comments can optionally be followed by a
+     * single uncommented line containing the names of the data columns.
```
uncommented = "commented out"? Maybe just say "a single line starting with '#'" to avoid the ambiguity

bangerth commented on source/utilities.cc:
```diff
@@ -878,6 +880,48 @@ namespace aspect
+      AssertThrow(column_position != data_component_names.end(),
+                  ExcMessage("There is no data column named " + column_name
+                             + " in the current data file. Please check the name and the "
+                             "first uncommented line of your data file."));
```
same as above -- the term "uncommented" is ambiguous

bangerth commented on source/utilities.cc:
```diff
@@ -921,33 +965,90 @@ namespace aspect
+      while (true)
+        {
+          AssertThrow (name_column_index < 100,
+                       ExcMessage("The program found more than 100 columns in the first line of the data file. "
```
:-)
